src/subxt_transaction.rs

```diff
@@ -496,10 +505,142 @@ pub(crate) fn build_eth_tx_payload(
+	// recreate the tx params in a loop when we can't create a partial tx with the online
+	// client, due to various RPC related issues or state not being up to date (currently we
+	// handle an error which happens when trying to create a partial tx that is based on a
+	// certain finalized block returned by the RPC, which is then reported as not found).
+	// Retrying seems to fix the issue.
+	fn tx_params<CC: subxt::Config>(
+		mortality: &Option<u64>,
+		nonce: u64,
+		recipe: &TransactionRecipe,
+	) -> <DefaultExtrinsicParams<CC> as ExtrinsicParams<CC>>::Params {
+		let mut params = <SubstrateExtrinsicParamsBuilder<CC>>::new().nonce(nonce).tip(recipe.tip);
+		if let Some(mortal) = mortality {
+			params = params.mortal(*mortal);
+		}
+		params.build()
+	}
+
+	async fn current_finalized_block_number<CC: subxt::Config, KEYP>(
+		sink: &SubxtTransactionsSink<CC, KEYP>,
+	) -> u64
+	where
+		KEYP: Signer<CC> + Clone + Send + Sync + 'static,
+		AccountIdOf<CC>: Send + Sync + AsRef<[u8]>,
+	{
+		let block_ref = sink
+			.api()
+			.backend()
+			.latest_finalized_block_ref()
+			.await
+			.expect("to get the last finalized block ref. qed");
+		let block = sink
+			.api()
+			.blocks()
+			.at(block_ref)
+			.await
+			.expect("to get the corresponding block header. qed");
+		block.number().into()
+	}
+
+	let params = tx_params(mortality, nonce as u64, recipe);
+	let tx_call = generate_payload(to_account_id.clone(), recipe);
+	match sink.api().tx().create_partial(&tx_call, from_account_id, params.into()).await {
+		Ok(mut tx) => {
+			let submittable_tx = tx.sign(from_keypair);
+			let hash = submittable_tx.hash();
+			let block_number = current_finalized_block_number(sink).await;
+			let tx = SubxtTransaction::<C>::new(
+				submittable_tx,
+				nonce,
+				mortality.map(|mortal| block_number + mortal),
+				sink.get_to_account_metadata(account).expect("account metadata exists"),
+			);
+			debug!(target:LOG_TARGET,"built mortal tx hash: {:?}", hash);
+			Ok(tx)
+		},
+		Err(err) => match err {
+			subxt::Error::Block(subxt::error::BlockError::NotFound(_)) => {
+				let tx_call = generate_payload(to_account_id.clone(), recipe);
+				for _ in 0..DEFAULT_RETRIES_FOR_PARTIAL_TX_CREATION {
+					let params = tx_params(mortality, nonce as u64, recipe);
+					match sink
+						.api()
+						.tx()
+						.create_partial(&tx_call, from_account_id, params.into())
+						.await
+					{
+						Ok(mut tx) => {
+							let block_number = current_finalized_block_number(sink).await;
+							let submittable_tx = tx.sign(from_keypair);
+							let hash = submittable_tx.hash();
+							let subxt_tx = SubxtTransaction::<C>::new(
+								submittable_tx,
+								nonce,
+								mortality.map(|mortal| block_number + mortal),
+								sink.get_to_account_metadata(account)
+									.expect("account metadata exists"),
+							);
+							debug!(target:LOG_TARGET,"built mortal tx hash: {:?}", hash);
+							return Ok(subxt_tx)
```

Comment:
This still looks like a copy of the code in line 576-591.
Could you move everything into the single loop, without prelude? Is it doable?